### ISSUE TITLE
Segfault in janus_ice_send_thread()

### ISSUE THREAD

**Issue by MichaelB76** [meetecho/janus-gateway#231]
Hi, I've had a Janus crash in ice.c janus_ice_send_thread() on a call to nice_agent_send() (line 2590 in the current version of the file). Here's a paste to the backtrace:

http://pastebin.com/Z4WzUJaa

Looks like a null pointer was referenced in nice_socket_send(). This occurred during an automated test run over several hours that exercises the WebRTC datachannel. Multiple connections were left up to Janus (about 10 I think) but there were no connections or disconnections happening at the time.

The problem may be timing related, we noticed that the Janus memory usage was climbing and in the log there are a number of warning messages like this:

[33m[WARN][0m [3871686646] Didn't receive audio for more than a second...

...which could be an indication that packets weren't being processed in a timely manner.

Please let me know if there is any more information I can provide.

**Comment by ploxiln** [comment]
quick tip: you can get more detailed backtraces if you install (your distro's equivalent of) libglib2.0-0-dbg and libnice-dbg

**Comment by MichaelB76** [comment]
@ploxiln - Thanks for the tip. I'll try installing this and see if I can reproduce the problem. I'll also search through the code and see if I can find any candidates that may have set one of the variables to null.

**Comment by MichaelB76** [comment]
I installed the debug symbols for libnice as suggested and have a little more info in this paste:

http://pastebin.com/QEDxD17a

I also included the output of the 'info threads' command in case that is useful.

**Comment by lminiero** [comment]
Not sure what's happening there. Does this happen when the related handle has been closed for some reason? Any related info on the Janus log before segfaulting?

Could you also elaborate on the increasing memory usage? Are you noticing many retransmissions, e.g., through the admin API? If it really is a consistent leak that maybe eventually leads to memory corruption, the only wait to check would be to rely on something like Valgrind (which would also greatly slow down Janus during the test, though).

**Comment by MichaelB76** [comment]
Hi. Looking again at the log there is a message towards the end which I'd missed:

Long poll time out for session 2835826791...

Looking at the code it doesn't appear that this timeout will close the session but is it possible that the last messages in the log have been lost when Janus crashed? If so then it's possible the session could have timed-out and been closed, which could make it related to issue #141. I'll have a think about how to go about testing this idea.

Also in the log I can see a few messages like this:

[33m[WARN][0m [3887891074] Didn't receive audio for more than a second...
[33m[WARN][0m [2510408496] Didn't receive audio for more than a second...
[33m[WARN][0m [3871686646] Didn't receive audio for more than a second...
[nua_i_notify]: 100 Trying
[3887891074] Notifying that we are NOT receiving audio
[3871686646] Notifying that we are NOT receiving audio
[2510408496] Notifying that we are NOT receiving audio

.. which could be an indication that Janus is running very slowly. I've re-run the test a couple of times and haven't been able to reproduce the crash. What happens is the memory usage increases until the sessions eventually time out and disconnect with this message:

Timeout expired for session 1140127903...

It was possible to reconnect after this point but the audio was breaking up a bit.

Concerning the memory leak I'm suspicious about the custom plugin we're using to evaluate Janus, so I'll investigate that first. I've already increased the timeout in the sessions watchdog from 5 to 60 seconds as a defensive measure. I'll let you know what I find.

**Comment by MichaelB76** [comment]
Just a quick update, I've fixed some memory leaks in our plugin that would have impacted performance but we haven't had a chance to re-test properly due to other work taking priority.

**Comment by lminiero** [comment]
Ok, keep us posted.

**Comment by lminiero** [comment]
Any update on this?

**Comment by MichaelB76** [comment]
I've finally got back to looking at this again this week. One thing I've found is missing calls to sdp_parser_free() in the SIP plugin, looking at the Sofia docs this should be called when the structure returned by sdp_parse() is finished with. I'll try and sort out a pull request for this.

I'm also seeing quite a few leaks coming from janus_push-event(), e.g:

==28827== 4,064 bytes in 2 blocks are still reachable in loss record 873 of 882
==28827==    at 0x4A08934: calloc (vg_replace_malloc.c:623)
==28827==    by 0x3C4824EEC6: g_malloc0 (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C4821AD36: ??? (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C482650F4: g_slice_alloc (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C48245345: g_list_prepend (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C48259680: g_queue_push_head (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C4821F326: g_async_queue_push_unlocked (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x3C4821F3AC: g_async_queue_push (in /usr/lib64/libglib-2.0.so.0.3800.2)
==28827==    by 0x43E19F: janus_push_event (janus.c:3531)
==28827==    by 0xAE652B2: sb_softclient_sip_handler (sb_softclient.c:2165)
==28827==    by 0x5121D14: nua_application_event (nua_stack.c:393)
==28827==    by 0x517ADE1: su_base_port_execute_msgs (su_base_port.c:280)
==28827==    by 0x517ADE1: su_base_port_getmsgs (su_base_port.c:202)

**Comment by lminiero** [comment]
> I've finally got back to looking at this again this week. One thing I've found is missing calls to sdp_parser_free() in the SIP plugin, looking at the Sofia docs this should be called when the structure returned by sdp_parse() is finished with. I'll try and sort out a pull request for this.

We already do that, e.g., see sdp.c:

https://github.com/meetecho/janus-gateway/blob/master/sdp.c#L46

This is also done when errors are catched. Not sure if there are scenarios where that may not be enough, though, or if it's a false allarm.

> I'm also seeing quite a few leaks coming from janus_push-event(), e.g:

The line 3531 in janus.c doesn't do anything... mae sure you're testing with an up-to-date version of Janus. Are you using WebSockets or HEST? If it's the former, we fixed a few things there, also with respect to last messages before closing a connection.

That said, I also noticed some leaks in notifications, but the queue we use for those is created with ```g_async_queue_new_full```, which means that when we unref it all its elements are freed automatically:

https://github.com/meetecho/janus-gateway/blob/master/janus.c#L405
https://github.com/meetecho/janus-gateway/blob/master/janus.c#L493
https://github.com/meetecho/janus-gateway/blob/master/janus.c#L5256

The queue is also automatically locked when use, so no race conditions should happen there. Unless a notification is created but not added (e.g., notification created after a session is over) and not properly deallocated after that?

That said, are those dumps done with Valgrind? If so, please try using AddressSanitizer, which should be much faster and not slow Janud down when using it, something that can "pollute" the results. You can find more info [here](https://janus.conf.meetecho.com/docs/debug).

**Comment by MichaelB76** [comment]
Ah yes, I've just found the calls to sdp_parser_free() in sdp.c thanks. However Valgrind is reporting several leaks for this, it seems to be caused by re-invite messages coming from the SIP client. The response handler (case nua-i_invite) is creating a new sdp_parser_t for each message, but not cleaning them up when the check on session->status breaks.

We're using the REST API. We've downloaded the latest janus code but I'm not running it yet as we have to update our client side code to connect to it.

Thanks, I'll have a look at AddressSanitizer.
